### ISSUE TITLE
vcs_tag() error output should be silenced if fallback is specified

### ISSUE THREAD

**Issue by mqudsi** [mesonbuild/meson#2240]
meson's `vcs_tag` function is very useful but the fact that it always generates error output if no annotated tag matching the current git revision is found is... unnecessary.

We use actual git revisions as our version numbers for intermediate releases, so we provide the output of `git rev-parse HEAD` as the value of the `fallback` parameter to `vcs_tag`. That works great, but each time meson complains about no annotated tags:

```
ninja -C debug
ninja: Entering directory `debug'
[1/4] Generating GitVersion.cpp with a custom command.
fatal: No annotated tags can describe 'e5c806c6a0fc0380637b955b018485ccd1214ddd'.
However, there were unannotated tags: try --tags.
```

If a `fallback` is provided, I think that output should be silenced.

**Comment by mqudsi** [comment]
(the fact that it says "fatal" when a fallback is provided is definitely a bug in all cases.)

**Comment by axxel** [comment]
Could you provide some sample meson.build code that shows what exactly you did and explain what you intended to achieve? I don't understand how/why you would 'provide the output of git...' as the fallback. The fallback is supposed to provide a string in case `git` could not be executed.

**Comment by mqudsi** [comment]
@axxel we are thinking of two different failure modes. The failure mode I was referring to is that there are no (matching?) annotated tags but `git` is executable.

**Comment by axxel** [comment]
Sorry but I still don't understand what you are trying to do. Can you provide the `vcs_tag`-line from your meson.build file?

**Comment by mqudsi** [comment]
I'm trying to use git (tag or revision) as the build tag.

Here's the excerpt from meson.build:
```python
tag_target = vcs_tag(command: ['git', 'rev-parse', 'HEAD'], input: 'utils/GitVersion.cpp', output: 'GitVersion.cpp')
```

and here's `utils/GitVersion.cpp`:
```cpp
#ifndef STRINGIFY
#define STRINGIFY2(x) #x
#define STRINGIFY(x) STRINGIFY2(x)
#endif

#include <string>

namespace neosmart
{
	//std::string GitVersion = STRINGIFY(TPVERSION);
	std::string GitVersion = "@VCS_TAG@";
}
```

**Comment by mqudsi** [comment]
Previously I was using the fallback tag:

```
rev = run_command('git', 'rev-parse', 'HEAD').stdout().strip()
tag_target = vcs_tag(input: 'utils/GitVersion.cpp', output: 'GitVersion.cpp', fallback: rev)
```

but I now switched to the `command` parameter.

**Comment by axxel** [comment]
I'm getting close but still don't understand it fully. Please note two things:
 1. your previous approach and your new approach are not equivalent, i.e. they produce different `GitVersion.cpp` files. Your new approach is the correct way. And it should actually work as is.
 1. the error message you see on the console seems to be from git, not from meson.
Unfortunately I don't know enough about git to interpret the error message. If I call `git rev-parse HEAD` in a tree, I simply get a proper hash code. Googling for the `fatal: ...` line hints at some missing history in your clone that you'd need to fetch if you wanted info about tags. But I fail to see how `rev-parse` would care about tags at all.

Do you still get this error text with your new approach? What happens if you simply call `git rev-parse HEAD` in your source tree?

**Comment by mqudsi** [comment]
meson internally calls `git describe` to generate the `vcs_tag()` output. `git describe` will throw an error if there are no tags (previously: if there were no annotated tags) that can describe the current revision.

When I filed the bug, I thought "No annotated tags can describe  ..." came from meson, not from git. In all cases, it did not come from the code explicitly executed in meson.vim.

The bug was that a fatal error message is generated when a fallback value is provided. An error should not be printed since a fallback means "use this in case of an error".

**Comment by axxel** [comment]
Thanks for the explanation. So if I got this correct, we have the following situation:
 1. Neither your old nor your new code really 'failed' to perform the replacement of the VCS_TAG string.
 1. Your old code did likely not what you intended but that was because you used the vcs_tag not as it was supposed to be use.
 1. Your new code will in fact not display this 'fatal:...' error because it does not execute `git describe`.
 1. It would have helped (you and me) if the output on the console was clearly attributed to git.
 1. It would have made the whole debugging of the issue way worse, if meson simply swallowed the error message from git as you suggested.
 1. Please also note that a fallback is always provided, its default value is meson.project_version().

I would propose the following changes to improve the situation:
 1. add some hint to the meson output where the error text came from in case there is one.
 1. rethink the vcs_tag default command used for git repos. A quick check revealed an easy solution: add the `--always` flag.

@jpakkane would you merge a PR with suggestion two (or simply add that `--always` flag yourself)?

**Comment by mqudsi** [comment]
Yup, that sounds about right. Sorry for how hard it was to describe, it's been awhile since I reported the issue. I also believe I changed my usage of `vcs_tag()` in between, but since forgot which was what.

(Note that I wasn't aware that meson used `git describe` internally until now, thanks for finding the `--always` flag, that does make things way easier.)

**Comment by ernestask** [comment]
Not everyone builds from git, by the way. In those cases, the fatal errors are a mere annoyance.

**Comment by pwnorbitals** [comment]
Do we have any progress on this issue ? @axxel's suggestions are interesting

**Comment by JamBalaya56562** [comment]
Resolved on current Meson — both halves of this are handled.

In [`mesonbuild/scripts/vcstagger.py`](https://github.com/mesonbuild/meson/blob/master/mesonbuild/scripts/vcstagger.py),
`config_vcs_tag` runs the VCS command with `stderr=subprocess.DEVNULL` and, on
*any* failure, silently substitutes the `fallback`:

    output = subprocess.check_output(cmd, cwd=source_dir, stderr=subprocess.DEVNULL)
    ...
    except Exception:
        new_string = fallback

So the `fatal: No annotated tags can describe ...` text (which, as established
later in this thread, comes from `git describe`, not Meson) no longer reaches the
console, and the `fallback` is used without noise.

On top of that, the default command is now `git describe --dirty=+ --always`
([`mesonbuild/utils/universal.py`](https://github.com/mesonbuild/meson/blob/master/mesonbuild/utils/universal.py)):
the `--always` flag (the one @axxel identified) makes `git describe` fall back to
the abbreviated commit hash instead of failing, so the default `vcs_tag()` path
doesn't error even without annotated tags. And because the `except` catches
everything, @ernestask's "not everyone builds from git" case also falls back
quietly instead of printing errors.

This looks resolved and can be closed as fixed. If anyone still sees stray VCS
error output on a recent release, please share a reproducer.
